vms/platformvm/validators/manager.go

```diff
@@ -141,137 +165,142 @@ func (m *manager) GetValidatorSet(ctx context.Context, height uint64, subnetID i
 			return nil, err
 		}
+		for nodeID, pk := range pkDiffs {
+			if vdr, ok := validatorSet[nodeID]; ok {
+				// The validator's public key was removed at this block, so it
+				// was in the validator set before.
+				vdr.PublicKey = pk
+			}
+		}
 	}
-
-	// cache the validator set
-	validatorSetsCache.Put(height, subnetSet)
-
-	endTime := m.clk.Time()
-	m.metrics.IncValidatorSetsCreated()
-	m.metrics.AddValidatorSetsDuration(endTime.Sub(startTime))
-	m.metrics.AddValidatorSetsHeightDiff(lastAcceptedHeight - height)
-	return subnetSet, nil
+	return validatorSet, nil
 }
 
-func (m *manager) rebuildPrimaryValidatorSet(
-	primarySet map[ids.NodeID]*validators.GetValidatorOutput,
-	height uint64,
-) error {
-	primaryWeightDiffs, err := m.state.GetValidatorWeightDiffs(height, constants.PlatformChainID)
-	if err != nil {
-		return err
-	}
-	for nodeID, weightDiff := range primaryWeightDiffs {
-		if err := applyWeightDiff(primarySet, nodeID, weightDiff); err != nil {
-			return err
+func (m *manager) makeSubnetValidatorSet(
+	currentHeight uint64,
+	targetHeight uint64,
+	subnetID ids.ID,
+) (map[ids.NodeID]*validators.GetValidatorOutput, error) {
+	currentValidators, ok := m.cfg.Validators.Get(subnetID)
+	if !ok {
+		currentValidators = validators.NewSet()
+		if err := m.state.ValidatorSet(subnetID, currentValidators); err != nil {
+			return nil, err
 		}
 	}
+	currentValidatorList := currentValidators.List()
 
-	pkDiffs, err := m.state.GetValidatorPublicKeyDiffs(height)
-	if err != nil {
-		return err
+	// Node ID --> Validator information for the node validating the Subnet.
+	subnetValidatorSet := make(map[ids.NodeID]*validators.GetValidatorOutput, len(currentValidatorList))
+	for _, vdr := range currentValidatorList {
+		subnetValidatorSet[vdr.NodeID] = &validators.GetValidatorOutput{
+			NodeID: vdr.NodeID,
+			// PublicKey will be picked from primary validators
+			Weight: vdr.Weight,
+		}
 	}
-	for nodeID, pk := range pkDiffs {
-		if vdr, ok := primarySet[nodeID]; ok {
-			// The validator's public key was removed at this block, so it
-			// was in the validator set before.
-			vdr.PublicKey = pk
+
+	// Rebuild subnet validators at [targetHeight]
+	for diffHeight := currentHeight; diffHeight > targetHeight; diffHeight-- {
+		weightDiffs, err := m.state.GetValidatorWeightDiffs(diffHeight, subnetID)
+		if err != nil {
+			return nil, err
+		}
+
+		for nodeID, weightDiff := range weightDiffs {
+			if err := applyWeightDiff(subnetValidatorSet, nodeID, weightDiff); err != nil {
+				return nil, err
+			}
 		}
 	}
-	return nil
-}
 
-func (m *manager) rebuildSubnetValidatorSet(
-	subnetSet, primarySet map[ids.NodeID]*validators.GetValidatorOutput,
-	subnetID ids.ID,
-	height uint64,
-) error {
-	targetWeightDiffs, err := m.state.GetValidatorWeightDiffs(height, subnetID)
+	// Get the public keys for all the validators at [targetHeight]
+	primarySet, err := m.getValidatorSetFrom(currentHeight, targetHeight, constants.PrimaryNetworkID)
```

Comment:
Note: we don't call `makePrimaryNetworkValidatorSet` because it may be cached. This does result in tracking another lookup in the metrics, but I feel like that's fine.